### ISSUE TITLE
Access API: Get All needs to remove services

### ISSUE THREAD

**Issue by junetaw** [on-api/on-api#47]
When getting all accesses from certain CO (especially larger ones) the response becomes very large. This is partly because they have a lot of accesses but more specifically that they tend to have a lot of services.

As an example from one CO a single access is over 12 000 characters minified with 100 services listed. Multiply with 100000+ accesses and everything stops working.

This will not scale, so my suggestion is that the "/accesses" endpoint DOES NOT contain services.
If there is a need to check services, then use the specific "/accesses/{accessID}".

If there IS a need to get all accesses AND have all services, then the API will probably need to implement a pageing structure to get batches of 100 or 1000 accesses.

**Comment by JonasEklundh** [comment]
Solution 1:  Paging
The API will always return a maximum of X posts, and information about paging. This solves two major resource hogs - memory needed for the API server to fetch, populate and serve JSON data, and memory for the client that needs to fetch an enormous JSON object, store it in memory and parse it. This does not change bandwidth usage and amount of data that needs to be transferred.

Solution 2: Separation
Remove services entirely from the accesses API and introduce a new API that lists services. The problem is data duplication, where there are - say - 200 services available for 50000 accesses in different forms. One way to handle this is that the API lists all services, with their basic information, and with support for deviations, example:

```
{
  "name": "B2100100",
  "connection": "2021-03-01",
  "available": "YES",
  "forcedTakeoverPossible": false,
  "comments": [
    "This service is limited to 100 Mbps for up to two weeks after ordering"
  ],
  "deviations": [
      {
        "field": "connection",
        "value": "2021-05-01",
        "accesses": ["987987", "98695", "07897", "898066"]
      },
      {
        "field": "connection",
        "value": "2021-06-01",
        "accesses": ["876876", "8687876", "i609860"]
      },
      {
        "field": "comment",
        "value": ["This service is included in a marketing campaign"],
        "accesses": ["68168168", "55716871", "58718671"]
      }
  ]
}

```

And move general information about an access connection and disconnection (not in relation to a specific service) to the accesses API. I.e:

```
{
  "accessId": "8732c2f065e2490babce820e94b1011a",
  "legacyAccessId": "KGXF49861",
  "streetName": "Testvägen",
  "streetNumber": "100",
  "streetLittera": "",
  "postalCode": "10000",
  "city": "Ankeborg",
  "countryCode": "SE",
  "premisesType": "MDU_APARTMENT",
  "mduApartmentNumber": "1001",
  "mduDistinguisher": "12121212",
  "outlet": "A-11-14",
  "population": "Hemsöhem",
  "accessState": "CONNECTED",
  "connection": "2021-03-05",
  "disconnection": false,
  "coNetworkAgreement": "NOTREQUIRED",
  "coFiberConverter": "LASER_3001X_MK2",
  "coCpeSwitch": "",
  "coCpeRouter": "NETGEAR WNDR4000"
}
```

So connection/disconnection is of the *access* and any deviation in relation to a specific service will be found in the /services API as above. This way we minimize the amount of data needed to convey the same information.

Disclaimer: I am not a consumer of the above API's and my comments are only in relation to my understanding of the use cases for the data found in them and ideas about how to make them more data-effective. I may have misunderstood the use and the lists of accesses in deviations would be hundreds of thousands of entries long as well. Feedback is appreciated.

**Comment by JonasEklundh** [comment]
Also adding a more versatile endpoint could be helpful, with `/services`, `/services/deviations`, `/services/deviations/connection` to cut down on the amount of data each request generates

**Comment by junetaw** [comment]
I think the proposed solution was to implement paging to resolve the more pressing problem.
But I like a "raw address" endpoint for import of address data.

**Comment by junetaw** [comment]
This issue is still alive and four feedback items from the current iteration (2.5.0) is related to this problem.

To keep the issue in line with the feedbacks, here is a short summary:

Address importing to SP:
Currently, SPs uses the access API, but that API is bloated with "services" which adds unneccecary time to the response time. The CO administration system must query and prepare the services response adding significant time and size to every access item. 

The proposed solutions are:
1. Add a new endpoint for /addresses (and optionally add webhooks for updates)
2. Remove the services sub-part of the response for the get all accesses operation (just as was done concerning subscriptions)
3. If neither 1 or 2 is possible then add an option in the request to include or exclude services.

**Comment by joeseos** [comment]
> This issue is still alive and four feedback items from the current iteration (2.5.0) is related to this problem.
> 
> To keep the issue in line with the feedbacks, here is a short summary:
> 
> Address importing to SP: Currently, SPs uses the access API, but that API is bloated with "services" which adds unneccecary time to the response time. The CO administration system must query and prepare the services response adding significant time and size to every access item.
> 
> The proposed solutions are:
> 
> 1. Add a new endpoint for /addresses (and optionally add webhooks for updates)
> 2. Remove the services sub-part of the response for the get all accesses operation (just as was done concerning subscriptions)
> 3. If neither 1 or 2 is possible then add an option in the request to include or exclude services.

I’ve heard that some service providers are caching the data from the /accesses endpoint to be able to increase the probability of placing a successfull order if the API goes down. However, maybe services and addresses could be separated with accessId as the common value to still be able to support the use case. 

And yes please, introduce webhooks, gRPC or similar :)

**Comment by junetaw** [comment]
The last suggestion, from Slack, was that /accesses, /services and /subscritions should be created/modified as follows;

/accesses 
 - Services sub-part is removed from access entity
 - Returns all accesses and intended use is to import deliverable accesses in an optimized structure. Minimal size and short response time is focus.

/accesses/{accessId}
 - Services sub-part is removed from access entity
 - Returns a single access

/services
 - Returns the services sub-part that was removed from access entity for all accesses,
 - Se example in the end.

/services/{accessId)
 - Returns the services sub-part that was removed from access entity for a single accesses,

/subscriptions
 - Returns all subscriptions (exists already today)

/subscriptions/{accessId)
 - Returns subscriptions for a single access (new)
 - Note: Issue #24 has other requests as well.


Service entity example response for get all:
[
  {
    "accessId": "8732c2f065e2490babce820e94b1011a",
    "services": [
      {
        "service": "BB-100-100",
        "connection": "2014-03-01",
        "available": "YES",
        "disconnection": ""
      },
      {
        "service": "BB-100-10",
        "connection": "2013-10-12",
        "available": "YES",
        "disconnection": ""
      },
      {
        "service": "BB-10-10",
        "connection": "2013-10-12",
        "available": "NO",
        "disconnection": "2018-12-31"
      },
      {
        "service": "IPTV",
        "connection": "2013-10-12",
        "available": "2019-01-01",
        "disconnection": ""
      },
      {
        "service": "VOIP",
        "connection": "NO",
        "available": "YES",
        "disconnection": ""
      },
    { ... next accessId.... }
    ]
  }
]

**Comment by joeseos** [comment]
@junetaw I like the suggestion, but I think we should add some information about what we're thinking about updates on /accesses and maybe even /services as some SPs like to cach this data, instead of last-modified-since, for example webhook, gRPC etc.

**Comment by junetaw** [comment]
@joeseos Yes, that would be good but isn't that a separate issue? They are not mutually exclusive. My aim is to make /accesses increadibly streamlined and quick, both when it comes to CPU cycles for the CO-system to get the response compiled and the size of the response. I have currently problems with a CO where the response seems to be greater than 2GB which is really troublesome for Windows based systems.

My hope is for Netadmin, Flow, BBE etc to be able to store a single file for their accesses that are immideately available cached. It of course is updated when they change an access, but otherwise no processing power should be necessary. We should be able to call on /accesses often as it should not therefore impact the CO systems.

If ISP needs to cache available services then the new /services endpoint is available and gives that data as before. As this endpoint probably causes a lot of processing (in order to determine which available services should be included with the access and if they are activatable etc) and is probably containing much more data it might be throttled by CO.

Yes, this is a breaking change but given a head start I do not think the ISPs should have trouble implementing this change.

**Comment by joeseos** [comment]
I agree :)

**Comment by junetaw** [comment]
A good point raised in #24 is that also subscription sub-part should be removed from /accesses since it moves to /subscription. It is only included when getting a single access (/accesses/{accessId}) anyway which is a little bit an anomaly.

**Comment by maxpaj** [comment]
Is there any progress on this?

**Comment by teemufrosen** [comment]
@maxpaj this has been implemented in 2.6.0 which is to be released this week.
